// Copyright 2020 The gRPC Authors

syntax = "proto3";

package envoy.service.discovery.v3;

import "envoy/config/core/v3/base.proto";

import "google/protobuf/any.proto";
import "google/protobuf/duration.proto";

message Status {

  int32 code = 1;

  string message = 2;

  repeated google.protobuf.Any details = 3;
}

message ResourceName {

  string name = 1;
}

message ResourceError {

  ResourceName resource_name = 1;

  Status error_detail = 2;
}

message DiscoveryRequest {

  string version_info = 1;

  config.core.v3.Node node = 2;

  repeated string resource_names = 3;

  string type_url = 4;

  string response_nonce = 5;

  Status error_detail = 6;
}

message DiscoveryResponse {

  string version_info = 1;

  repeated google.protobuf.Any resources = 2;

  bool canary = 3;

  string type_url = 4;

  string nonce = 5;

  repeated ResourceError resource_errors = 7;
}

message Resource {

  message CacheControl {

    bool do_not_cache = 1;
  }

  string name = 3;

  repeated string aliases = 4;

  string version = 1;

  google.protobuf.Any resource = 2;

  google.protobuf.Duration ttl = 6;

  CacheControl cache_control = 7;
}
